src/generator.cs

```
@@ -4850,6 +4850,23 @@ void GenerateAsyncMethod (MemberInformation original_minfo, AsyncMethodKind asyn
 		var mi = original_minfo.method;
 		var minfo = new AsyncMethodInfo (this, original_minfo.type, mi, original_minfo.category_extension_type, original_minfo.is_extension_method);
 		var is_void = mi.ReturnType == TypeManager.System_Void;
+
+		// Print a error if any of the method parameters or handler parameters is ref/out, it should not be asyncified.
+		if (minfo.async_initial_params != null) {
+			foreach (var param in minfo.async_initial_params) {
+				if (param.ParameterType.IsByRef) {
+					ErrorHelper.Show (new BindingException (1062, true, $"The '{original_minfo.type.Name}.{mi.Name}' member contains ref/out parameters and must not be decorated with [Async]."));
```

Minor: you can just throw the exception instead of calling `ErrorHelper.Show`:

```csharp
throw new BindingException (...);
```